test/support/generators.ex

```diff
@@ -0,0 +1,293 @@
+       * e.g. `["null", "integer", "boolean"]`
+    * `:excluding` - a list of `Schematic.kind` values that determines the **only** data types that **should not** be generated. See the `Schematic` module for available `kind`s. This option is mutually exclusive with `:using`; if the `:using` option is **not** an empty list then this option is ignored. Defaults to `[]`. 
+       * e.g. `["null", "integer", "boolean"]`
+  """
+  def schematic_and_data(opts \\ []) do
+    StreamData.bind(data(0, opts), fn raw_data ->
+      StreamData.constant({schematic_from_data(raw_data) |> Enum.fetch!(1), raw_data})
+    end)
+  end
+
+  defp data(depth, opts \\ [])
+
+  defp data(8, opts) do
+    using = Keyword.get(opts, :using, [])
+    excluding = Keyword.get(opts, :excluding, [])
+
+    filter =
+      case using do
+        [] -> {:reject, excluding}
+        _ -> {:accept, using}
+      end
+
+    ["integer", "boolean", "null", "string", "list", "tuple", "map"]
+    |> then(fn kinds ->
+      case filter do
+        {:reject, excluding} -> Enum.reject(kinds, &Enum.member?(excluding, &1))
+        {:accept, using} -> Enum.filter(kinds, &Enum.member?(using, &1))
+      end
+    end)
+    |> Enum.map(fn
+      "integer" ->
+        StreamData.integer()
+
+      "boolean" ->
+        StreamData.boolean()
+
+      "null" ->
+        StreamData.constant(nil)
+
+      "string" ->
+        StreamData.binary()
+
+      "tuple" ->
+        StreamData.list_of(scalar(), max_length: 5) |> StreamData.map(&List.to_tuple/1)
+
+      "list" ->
+        StreamData.list_of(scalar(), max_length: 5)
+
+      "map" ->
+        StreamData.map_of(
+          StreamData.one_of([StreamData.binary(), StreamData.integer()]),
+          scalar(),
+          max_length: 5
+        )
+    end)
+    |> StreamData.one_of()
+  end
+
+  defp data(depth, opts) do
+    using = Keyword.get(opts, :using, [])
+    excluding = Keyword.get(opts, :excluding, [])
+
+    filter =
+      case using do
+        [] -> {:reject, excluding}
+        _ -> {:accept, using}
+      end
+
+    ["integer", "boolean", "null", "string", "list", "tuple", "map"]
+    |> then(fn kinds ->
+      case filter do
+        {:reject, excluding} -> Enum.reject(kinds, &Enum.member?(excluding, &1))
+        {:accept, using} -> Enum.filter(kinds, &Enum.member?(using, &1))
+      end
+    end)
+    |> Enum.map(fn
+      "integer" ->
+        StreamData.integer()
+
+      "boolean" ->
+        StreamData.boolean()
+
+      "null" ->
+        StreamData.constant(nil)
+
+      "string" ->
+        StreamData.binary()
+
+      "tuple" ->
+        StreamData.list_of(data(depth + 1), max_length: 5) |> StreamData.map(&List.to_tuple/1)
+
+      "list" ->
+        StreamData.list_of(data(depth + 1), max_length: 5)
+
+      "map" ->
+        StreamData.map_of(
+          StreamData.one_of([StreamData.binary(), StreamData.integer()]),
+          data(depth + 1),
```

Comment:
actually, i would also remove one of these function clauses and change this line to. too much code duplication for the one line of change that halts the recursion
```suggestion
          case depth + 1 do
            8 -> scalar()
            depth -> data(depth)
          end,
```

Reply:
okay. pretty sure i understand what you're saying. i'll make the change

Reply:
it's _much_ better after those changes, thanks for the suggestion

Reply:
Excelente